Feature: SSP Tables (#925)
LLNL/sundials: doc/arkode/guide/source/Butcher.rst

Reviewer: Update the tables at the start of the ERK and DRIK sections with the new methods

Reviewer: [This comment](https://github.com/llnl/sundials/pull/925/changes#r3295338061) wasn't clear. I was referring to this table and the similar one below for DIRK methods that need to be udpated.

Reviewer: [This comment](https://github.com/llnl/sundials/pull/925/changes#r3295338061) wasn't clear. I was referring to this table and the similar one above for ERK methods.

     1 & 0 & \frac35 & \frac25 \\
   \end{array}

where :math:`\gamma = \frac{2-\sqrt{2}}{2}`.

.. figure:: /figs/arkode/ARKODE_ASCHER_SDIRK_3_1_2_stab_region.png
   :width: 50 %
   :align: center

   Linear stability region for the Ascher-SDIRK-3-1-2 method. The method's
   region is outlined in red; the embedding's region is in blue dashed.

.. c:enumerator:: ARKODE_SSP_DIRK_3_1_2

Accessible via the constant ``ARKODE_SSP_DIRK_3_1_2`` to
:c:func:`ARKStepSetTableNum`, or
:c:func:`ARKodeButcherTable_LoadDIRK`.
Accessible via the string ``"ARKODE_SSP_DIRK_3_1_2"`` to
:c:func:`ARKStepSetTableName`, or
:c:func:`ARKodeButcherTable_LoadDIRKByName`.
The original method is the "IMEX-SSP2(3,3,2)" stiffly accurate method from
:cite:p:`PR:2005`, with an embedding invented specifically for ARKODE.  The method
is both A- and L-stable, while the embedding is A-stable.  Both the method and
embedding have SSP coefficient equal to 2.4.

.. math::

   \renewcommand{\arraystretch}{1.5}
   \begin{array}{r|ccc}
     \frac14 & \frac14 & 0       & 0 \\
     \frac14 & 0       & \frac14 & 0 \\
     1       & \frac13 & \frac13 & \frac13 \\
     \hline
     2 & \frac13 & \frac13 & \frac13 \\

>      1 & \frac13-\frac{1}{10} & \frac13 & \frac13+\frac{1}{10} \\

Reviewer 2: ```suggestion
     1 & \frac{7}{30} & \frac13 & \frac{13}{30} \\
```

.. math::

   \renewcommand{\arraystretch}{1.5}
   \begin{array}{r|ccc}
     \frac{2}{11}    & \frac{2}{11}      & 0 & 0 \\
     \frac{289}{462} & \frac{205}{462}   & \frac{2}{11}   & 0 \\
     \frac{751}{924} & \frac{2033}{4620} & \frac{21}{110} & \frac{2}{11} \\
     \hline
     2 & \frac{24}{55} & \frac{1}{5} & \frac{4}{11} \\
     1 & \frac{4247}{9840} & \frac{21343}{108240} & \frac{49}{132}
   \end{array}

.. figure:: /figs/arkode/ARKODE_SSP_LSPUM_SDIRK_3_1_2_stab_region.png
   :width: 50 %
   :align: center

   Linear stability region for the SSP-LSPUM-SDIRK-3-1-2 method.  The method's
   region is outlined in red; the embedding's region is in blue dashed.

.. c:enumerator:: ARKODE_SSP_ESDIRK_4_2_3

Accessible via the constant ``ARKODE_SSP_ESDIRK_4_2_3`` to
:c:func:`ARKStepSetTableNum` or :c:func:`ARKodeButcherTable_LoadDIRK`.
Accessible via the string ``"ARKODE_SSP_ESDIRK_4_2_3"`` to
:c:func:`ARKStepSetTableName` or
:c:func:`ARKodeButcherTable_LoadDIRKByName`.
This method was invented specifically for ARKODE as an implicit method to
pair with ``ARKODE_SSP_ERK_4_2_3`` for an embedded additive
Runge--Kutta method.  The method is both A- and L-stable, and the

> embedding is A-stable.  Neither the ESDIRK method or its embedding are SSP.

Reviewer: If the method is not SSP should we drop that from the name?

Author: That's a good question.  I'd named it this way so that it was obviously part of the ARK pair along with `ARKODE_SSP_ERK_4_2_3`, but perhaps we don't need the names to align?

Reviewer: I don't think the names need align since having SSP in the name of a non-SSP method would be confusing and we have a table that lists which methods form an ImEx pair. 

Side note: At some point, I think we should add a new type for ImEx methods that holds both tables and users can load the pair by an ImEx name or ID without having to know both table names.

analysis considers the additive Dahlquist test problem

.. math::

   \dot{y}(t) = \lambda^I y + \lambda^E y, \quad y(0) = 1,

where the additive Runge--Kutta method applies the partition
:math:`f^I(t,y) = \lambda^I y` and :math:`f^E(t,y) = \lambda^E y`.  Using a
step size of :math:`h`, with corresponding scaled values
:math:`\eta^I = h\lambda^I` and :math:`\eta^E = h\lambda^E`, this gives rise
to the ARK stability function

.. math::

   R(\eta^I,\eta^E) = 1 + \left(\eta^I b^I + \eta^E b^E\right) [I - \eta^I A^I - \eta^E A^E]^{-1} e.

Then for a given angle :math:`\theta \in \left[0,\frac{\pi}{2}\right]` and radius :math:`\rho>0`, we may define the joint linear stability region

.. math::

   \mathcal{J}_{\theta,\rho} = \left\{ \eta^E \in \mathbb{C}\; :\; \left| R(\eta^I,\eta^E) \right| \le 1,\; \forall \eta^I\in S(\theta,\rho)\right\},

where the sector :math:`S(\theta,\rho)` is defined as

.. math::

   S(\theta,\rho) = \left\{ \eta^I = -a+i b \; :\; a>0,\; b\ge 0,\; |\eta^I|\le \rho,\; \operatorname{atan}\left(\frac{b}{a}\right) \le \theta \right\}.

For each ARK method, we overlay plots of the border of :math:`\mathcal{J}_{\theta,\rho}`

> for values :math:`\theta \in \left\{0^o, 20^0, 40^0, 60^0, 80^0\right\}` and

Reviewer 2: Can you use degree symbol?

> for values :math:`\theta \in \left\{0^o, 20^0, 40^0, 60^0, 80^0\right\}` and

Reviewer 2: I think the most useful angle to include would be 90

:math:`\rho=10^8`; roughly corresponding with the shape of the explicit stability region

> under the assumption that the implicit method is :math:`A-\theta` stable;

Reviewer: ```suggestion
under the assumption that the implicit method is :math:`A(\theta)`-stable;
```

Reviewer: `A-\theta` looks too much like `A minus \theta` which is probably why `A(\theta)` is more commonly used (e.g., in Hairer and Wanner).

> the joint stability region decreases in area as :math:`\theta` and :math:`\rho` increase.

Reviewer 2: ```suggestion
under the assumption that the implicit method is :math:`A-\theta` stable;
the joint stability region decreases in area as :math:`\theta` and :math:`\rho` increase.
```
This is always the case

analysis considers the additive Dahlquist test problem

.. math::

   \dot{y}(t) = \lambda^I y + \lambda^E y, \quad y(0) = 1,

where the additive Runge--Kutta method applies the partition
:math:`f^I(t,y) = \lambda^I y` and :math:`f^E(t,y) = \lambda^E y`.  Using a
step size of :math:`h`, with corresponding scaled values
:math:`\eta^I = h\lambda^I` and :math:`\eta^E = h\lambda^E`, this gives rise
to the ARK stability function

.. math::

   R(\eta^I,\eta^E) = 1 + \left(\eta^I b^I + \eta^E b^E\right) [I - \eta^I A^I - \eta^E A^E]^{-1} e.

Then for a given angle :math:`\theta \in \left[0,\frac{\pi}{2}\right]` and radius :math:`\rho>0`, we may define the joint linear stability region

.. math::

   \mathcal{J}_{\theta,\rho} = \left\{ \eta^E \in \mathbb{C}\; :\; \left| R(\eta^I,\eta^E) \right| \le 1,\; \forall \eta^I\in S(\theta,\rho)\right\},

where the sector :math:`S(\theta,\rho)` is defined as

.. math::

   S(\theta,\rho) = \left\{ \eta^I = -a+i b \; :\; a>0,\; b\ge 0,\; |\eta^I|\le \rho,\; \operatorname{atan}\left(\frac{b}{a}\right) \le \theta \right\}.

For each ARK method, we overlay plots of the border of :math:`\mathcal{J}_{\theta,\rho}`

> for values :math:`\theta \in \left\{0^{\circ}, 20^{\circ}, 40^{\circ}, 60^{\circ}, 80^{\circ}\right\}` and

Reviewer: Align with plot legends
```suggestion
for values :math:`\theta \in \left\{0^{\circ}, 15^{\circ}, 30^{\circ}, 45^{\circ}, 60^{\circ}, 75^{\circ}, 90^{\circ}\right\}` and
```

:c:enumerator:`ARKODE_SSP_LSPUM_SDIRK_3_1_2`.

.. figure:: /figs/arkode/ARKODE_SSP_LSPUM_ARK_3_1_2_joint_stab_region.png
   :width: 50 %
   :align: center

   Joint linear stability regions :math:`\mathcal{J}_{\theta,10^8}` for the
   ``ARKODE_SSP_LSPUM_ARK_3_1_2`` method.

.. _Butcher.ARKODE_ARK324L2SA_ARK_4_2_3:

ARKODE_ARK324L2SA_ARK_4_2_3
^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. index:: ARKODE_ARK324L2SA_ARK_4_2_3 ARK method

3rd-order method that combines :c:enumerator:`ARKODE_ARK324L2SA_ERK_4_2_3`
with :c:enumerator:`ARKODE_ARK324L2SA_DIRK_4_2_3`.
This is the default third-order ARK method.

.. figure:: /figs/arkode/ARKODE_ARK324L2SA_4_2_3_joint_stab_region.png
   :width: 50 %
   :align: center

   Joint linear stability regions :math:`\mathcal{J}_{\theta,10^8}` for the
   ``ARKODE_ARK324L2SA_ARK_4_2_3`` method.  The joint stability regions only
   differ near the imaginary axis, so for most of the region only the
   :math:`80^o` boundary is visible.

.. _Butcher.ARKODE_SSP_ARK_4_2_3:

ARKODE_SSP_ARK_4_2_3
^^^^^^^^^^^^^^^^^^^^^^^^^^

.. index:: ARKODE_SSP_ARK_4_2_3 ARK method

3rd-order method that combines :c:enumerator:`ARKODE_SSP_ERK_4_2_3` with

> :c:enumerator:`ARKODE_SSP_ESDIRK_4_2_3`.

Reviewer: Since both tables are not SSP, should we drop SSP from the name?

Author: It is not a "settled" question whether SSP is in fact necessary for the implicit part of an ARK table.  In Pareschi and Russo 2005 (https://doi.org/10.1007/BF02728986), they point out that if the implicit table is just L-stable and the explicit table is SSP, then in the stiff limit the SSP-ness of the explicit method is obtained.  Later in 2014, Higueras (https://doi.org/10.1016/j.cam.2014.05.011) analyzed a separate "uniform convergence" property that requires both the explicit and implicit tables to be SSP (along with another property).  In my view, the literature on ImEx SSP methods is still too sparse to conclusively settle this topic.

My name for this table derives from the Pareschi & Russo view.

Reviewer: That's good with me. Especially if we drop SSP from the implicit table name so it is clear only the ERK is SSP.

   +----------------------------------------------+-----------------------------------------------+--------+----------------+-------+

.. _Butcher.ARKODE_ARK2_3_1_2:

ARKODE_ARK2_3_1_2
^^^^^^^^^^^^^^^^^

.. index:: ARKODE_ARK2_3_1_2 ARK method

2nd-order method that combines :c:enumerator:`ARKODE_ARK2_ERK_3_1_2` with
:c:enumerator:`ARKODE_ARK2_DIRK_3_1_2`.  This is the default
second-order ARK method.

.. figure:: /figs/arkode/ARKODE_ARK2_3_1_2_joint_stab_region.png
   :width: 50 %
   :align: center

   Joint linear stability regions :math:`\mathcal{J}_{\theta,10^8}` for the
   ``ARKODE_ARK2_3_1_2`` method.  All

>    joint stability regions coincide, so only the :math:`80^o` boundary is visible.

Reviewer: Update to match legend
```suggestion
   joint stability regions coincide, so only the :math:`90^{\circ}` boundary is visible.
```

.. _Butcher.ARKODE_ASCHER_ARK_3_1_2:

ARKODE_ASCHER_ARK_3_1_2
^^^^^^^^^^^^^^^^^^^^^^^

.. index:: ARKODE_ASCHER_ARK_3_1_2 ARK method

2nd-order method that combines :c:enumerator:`ARKODE_ASCHER_ERK_3_1_2` with
:c:enumerator:`ARKODE_ASCHER_SDIRK_3_1_2`.

.. figure:: /figs/arkode/ARKODE_ASCHER_ARK_3_1_2_joint_stab_region.png
   :width: 50 %
   :align: center

   Joint linear stability regions :math:`\mathcal{J}_{\theta,10^8}` for the
   ``ARKODE_ASCHER_ARK_3_1_2`` method.

>    All joint stability regions coincide, so only the :math:`80^o` boundary

Reviewer: Update to match legend
```suggestion
   All joint stability regions coincide, so only the :math:`90^{\circ}` boundary
```
